receiver/filereceiver/file.go

```diff
@@ -0,0 +1,147 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package filereceiver // import "github.com/open-telemetry/opentelemetry-collector-contrib/receiver/filereceiver"
+
+import (
+	"bufio"
+	"context"
+	"os"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/config"
+	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/model/otlp"
+	"go.opentelemetry.io/collector/receiver/receiverhelper"
+	"go.uber.org/multierr"
+)
+
+const (
+	typeStr = "file"
+)
+
+var (
+	tracesUnmarshaler  = otlp.NewJSONTracesUnmarshaler()
+	logsUmarshaler     = otlp.NewJSONLogsUnmarshaler()
+	metricsUnmarshaler = otlp.NewJSONMetricsUnmarshaler()
+)
+
+// NewFactory creates a factory for file receiver
+func NewFactory() component.ReceiverFactory {
+	return receiverhelper.NewFactory(
+		typeStr,
+		createDefaultConfig,
+		receiverhelper.WithMetrics(createMetricsReceiver),
+		receiverhelper.WithLogs(createLogsReceiver),
+		receiverhelper.WithTraces(createTracesReceiver))
+}
+
+// ReceiverType implements file.LogReceiverType
+// to create a file receiver
+type ReceiverType struct{}
+
+// Type is the receiver type
+func (f ReceiverType) Type() config.Type {
+	return typeStr
+}
+
+type cfg struct {
+	config.ReceiverSettings `mapstructure:",squash"` // squash ensures fields are correctly decoded in embedded struct
+	Path                    string                   `mapstructure:"path"`
+}
+
+func createDefaultConfig() config.Receiver {
+	return &cfg{
+		Path:             "",
+		ReceiverSettings: config.NewReceiverSettings(config.NewComponentID(typeStr)),
+	}
+}
+
+type receiver struct {
+	watcher watcher
+}
+
+func (f *receiver) Start(ctx context.Context, host component.Host) error {
+	return f.watcher.start()
+}
+
+func (f *receiver) Shutdown(ctx context.Context) error {
+	return f.watcher.stop()
+}
+
+func createLogsReceiver(_ context.Context, settings component.ReceiverCreateSettings, configuration config.Receiver, logs consumer.Logs) (component.LogsReceiver, error) {
+	readLogs := func(path string) error {
+		fileHandle, _ := os.Open(path)
```

Comment:
I am not a fan of lambdas within a function since it make it harder to provide debug information back to the user and being able to unit test them in isolation is problematic due to being coupled with a method that may not allow for unit testing.

Reply:
True, that's a miss. I can rework the lambda as a separate method, sure thing.

As for the file being opened by something else, etc, there's no guarantee in anything here. The file can be deleted, chmoded, moved, truncated, appended. It's going to be exactly as good as the operator makes it. This particular filereceiver is not very smart and we can spell that out in the docs if you like.

For concurrent access and live updates, we can work on kafka/redis/<your tech here> queues.

Reply:
Is it right to assume that an exporter writing to the listened folder will be able to atomically write to it ?

Reply:
No, I'm not particularly trying to be good about it right now.